proposals/2024-09.livemd

```diff
@@ -0,0 +1,96 @@
+# kubereq API improvements
+
+```elixir
+Mix.install([
+  {:kubereq, github: "mruoss/kubereq", ref: "compile-time-resource-path-lookup"},
+  {:kino, "~> 0.14.1"}
+])
+```
+
+## Motivation
+
+Currently, the `kubereq` library is used as follows:
+
+```elixir
+kubeconfig = Kubereq.Kubeconfig.load(Kubereq.Kubeconfig.Default)
+sa_req = Kubereq.new(kubeconfig, "api/v1/namespaces/:namespace/serviceaccounts/:name")
+
+Kubereq.get(sa_req, "default", "default") |> Kino.Tree.new()
+```
+
+I wrote `kubereq` and `kubegen` a few months ago but never got around to really use it. Now that I have used `kubereq` for the K8s runtime in Livebook, I feel like the API needs some minor adaptions.
+
+One part that calls for abstraction is loading the Kubeconfig. Most people probably get by with the default pipeline so while still allowing for explicit override, we can set it as default.
+
+Also, line 2 in the code above is a problem to me. I can never remember this path and always had to look it up or copy it. And I know Kubernetes quite well. For somebody with less deep knowledge of the API behind `kubectl`, it's probably not super straight forward to find out what is required.
+
+## Proposal 1: Attach function with kubeconfig as option
+
+We can offer `Kubereq.attach/2` where the second argument is an options Keyword list. The only option at the moment is `:kubeconfig`. If it is not defined, the default config `Kubereq.Kubeconfig.Default` is loaded. Otherwise, the user can pass their own pipeline or an already loaded `%Kubereq.Kubeconfig{}` struct.
+
+```elixir
+# loads Kubereq.Kubeconfig.Default implicitely:
+req1 = Req.new() |> Kubereq.attach()
+
+# pass Kubeconfnig pipeline which will be loaded by `Kubereq.attach/2`:
+req2 =
+  Req.new()
+  |> Kubereq.attach(
+    kubeconfig: {Kubereq.Kubeconfig.File, path: ".kube/config", relative_to_home?: true}
```

Comment:
Having kubeconfig as an option with the common case default sounds good, it could just as well be an option to `Kubereq.new`

While at it, is there a reason for using pluggable? (which I don't mind as an abstraction, it's just that it's an additional dependency) For example, here we could do this:

```elixir
Kubereq.Kubeconfig.from_file(path: ".kube/config", relative_to_home?: true)
```

If the user wants multiple alternatives, it's a simple chain logic they can do. But in practice, the default can try the common locations, and if the user wants a specific kubeconfig, they will likely load from one specific place.

Reply:
Yeah... good point. Here, pluggable is probably an overkill indeed. I have written it for [bonny](https://github.com/coryodaniel/bonny) where I think it actually does bring value but probably not so much here.

Going to address this separately, though.

Reply:
Oh and I'm using `relative_to_home` in `Kubereq.Kubeconfig.Default`. At compile time I don't know the user's home...

Reply:
I think you can allow users use ~, keep it as is at compile time and Path.expand it at runtime. I haven’t looked closely at that code so I might be missing something though!

Reply:
I'd like [`Kubereq.Kubeconfig.Default`](https://github.com/mruoss/kubereq/blob/main/lib/kubereq/kubeconfig/default.ex) to work whether on Windows or on Unix/Mac. But afaik, on Windows, `~` doesn't work, right?

Reply:
Ooh `Path.expand` resolves `~` no matter what!! Yeah I got it now! 😅